crates/matrix-sdk/src/widget/client/implementation/outgoing.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A high-level API to generate commands (requests) that we send either
+//! directly to the widget, or to the matrix driver/client.
+
+use ruma::{
+    api::client::account::request_openid_token::v3::Response as RumaOpenIdResponse,
+    events::AnyTimelineEvent, serde::Raw, OwnedEventId,
+};
+
+use super::{
+    super::actions::{ReadEventCommand, SendEventCommand},
+    openid::OpenIdResponse,
+};
+use crate::widget::Permissions;
+
+#[async_trait::async_trait]
+pub trait RequestSender {
+    async fn send<T: Request>(&self, request: T) -> Result<T::Response, String>;
+}
+
+pub trait Request {
+    type Response;
+}
+
+pub struct RequestPermissions;
```

Comment:
All these request structs also need docs, and should probably be `pub(crate)` to make it clear at a glance that it's not part of the overall crate's public API.

Reply:
The whole API is internal to the crate (in fact it's internal to the  `widget` module) and never exposed to the outside.

But I see your point, I added `#![warn(unreachable_pub)] to the `client/mod.rs` (consequently adding `pub(crate)` to each related part; I wonder if there is a way to avoid the boilerplate to mention `crate` everywhere, such as that one can essentially say "from this module downwards, nothing gets exposed").